### ISSUE TITLE
Clang with -std=c++2a crashes on invalid recursive variadic template struct definition

### ISSUE THREAD

**Issue by rgreenblatt** [llvm/llvm-project#181]
This happens with 10 rc2/rc3 and latest dev built from source.

This is a pretty minimal reproduction:
```
template <char First, char... Rest> struct Recursive {
  using next = typename Recursive<Rest...>::type;
  using type = notdefined<next::anything>;
};
```

Build with `clang++ -std=c++2a test.cpp`
It doesn't crash with `-std=c++17`

I made a [docker container which reproduces the issue](https://hub.docker.com/r/greenblattryan/clang-crash-recursive-template).

[backtrace.txt](https://github.com/llvm/llvm-project/files/4333084/backtrace.txt)

Here is a zip with the backtrace, preprocessed source and associated run script.
[bug-report.zip](https://github.com/llvm/llvm-project/files/4333081/bug-report.zip)

**Comment by rgreenblatt** [comment]
I ran into this when including some parts of libstdc++ and having errors from elsewhere. The fact that the struct in this case is invalid may not be the root cause of the issue.

**Comment by zmodem** [comment]
Bisection points to this commit:

---
commit 67c608a9695496cfc9d3fdf9d0b12b554ac6b4df                        
Author: Saar Raz <saar@raz.email>                                                              
Date:   Fri Jan 24 00:43:22 2020 +0200                                                         
                                                                                               
    [Concepts] Deprecate -fconcepts-ts, enable Concepts under -std=c++2a  
                                                                                               
    Now with concepts support merged and mostly complete, we do not need -fconcepts-ts
    (which was also misleading as we were not implementing the TS) and can enable
    concepts features under C++2a. A warning will be generated if users still attempt
    to use -fconcepts-ts.
---

@saarraz can you take a look?

**Comment by saarraz** [comment]
Will do in a few hours

**Comment by saarraz** [comment]
Fixed 19fccc52ff2c1da1f93d9317c34769bd9bab8ac8

**Comment by rgreenblatt** [comment]
This seems to be resolved, and I can't get any more crashes when using trunk.
